### PR TITLE
Merge 2.12.x to 2.13.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
```
$ git -P log $MB..origin/2.12.x --decorate --oneline --graph
```

// PICK
```
*   2ee5f6c783 (retronym/2.12.x, origin/2.12.x) Merge pull request #9500 from retronym/cleanup-cleanup
|\
| * f4da037f6b (retronym/cleanup-cleanup, origin/pr/9500) Fix non-exhaustive match in Cleanup phase
```

// SKIP (backport / 2.12.x specific)
```
*   eab3ce8693 Merge pull request #9507 from regadas/12344_bt_mima_report
|\
| * dc809a5649 (origin/pr/9507) Add missing MiMa filters for NewRedBlackTree
| * 0e0fb08da4 Fix mimaReportBinaryIssues, not run aggregate
|/
```

// PICK
```
* 01a8c458ed Merge pull request #9506 from retronym/topic/box-unbox-fix
* b705e3b467 (retronym/topic/box-unbox-fix, origin/pr/9506) Fix regression in box-unbox (in -opt:l:method)
```

// SKIP, backport
```
*   	 (origin/2.12.x) Merge pull request #9498 from bplommer/2.12.x
|\
| * 4a5d2b6dd2 Don't warnUnusedNowarn in scaladoc
```

// PICK
```
* |   83384b5086 Merge pull request #9499 from retronym/singleton-type-tree-pos
|\ \
| |/
|/|
| * a0f9840bfc (retronym/singleton-type-tree-pos, singleton-type-tree-pos) Retain positions of SingletonTypeTree
|/
```

// SKIP, already applied to 2.13
```
*   7bfcf50ede (origin/2.12.x) Merge pull request #9488 from SethTisue/asm-9.1
|\
| * d032ef8963 (origin/pr/9488) upgrade to asm 9.1, for JDK 17 support
|/
```

// PICK

```
*   f39a74abf0 Merge pull request #9475 from retronym/topic/stock-url-classloader
|\
| * 1e1a0fd6a9 (retronym/topic/stock-url-classloader, origin/pr/9475, topic/stock-url-classloader) Use a stock URLClassLoader in the scala runner for parallel capability
```

// SKIP, already applied 2.13.x
```
* |   3667278e85 Merge pull request #9477 from SethTisue/asm-9
|\ \
| * | 2a60c3121a (origin/pr/9477) upgrade to asm 9.0, for JDK 16 support
```

// PICK
```
* | |   adc9a22ef4 Merge pull request #9478 from retronym/ticket/12038-cyclic-found-reflection
|\ \ \
| |/ /
|/| |
| * | 845cda33e1 (retronym/ticket/12038-cyclic-found-reflection, origin/pr/9478, ticket/12038-cyclic-found-reflection) Fix cylic error in runtime reflection (protobuf)
|/ /
* |   5f6e5a1b5e (retronym/2.12.x, origin/pr/9469) Merge pull request #9461 from retronym/spire-array-regression
|\ \
| * | c9692130ae (retronym/spire-array-regression, origin/pr/9461, spire-array-regression) Faster `Array[T](t)` for @specialialized T.
| * | 3393616658 Fix regression in specialization of genericWrapArray(Array(..))
|/ /
```

// SKIP 2.12.x only
```
* |   a9b6d67e37 Merge pull request #9314 from mkeskells/2.12.x_RedBlackTree_serialisationAllocation
|\ \
| * | 646e6b7adb (origin/pr/9314) less allocation on deserialisation of RedBlackTree (SortedMap) reuse deserialisation logic for map and set
```

// PICK. 7d81f22f4e had non-trivial conflicts.
```
* | |   43a5a75bb1 Merge pull request #9082 from retronym/ticket/scala-async-152
|\ \ \
| * | | e309433964 (retronym/ticket/scala-async-152, origin/pr/9082, ticket/scala-async-152) async: await under synchronized is an error
* | | |   7ca86b97d5 Merge pull request #9455 from retronym/ticket/sd-741-pipeline-main-race
|\ \ \ \
| * | | | cf0b02796b (retronym/ticket/sd-741-pipeline-main-race, origin/pr/9455, ticket/sd-741-pipeline-main-race) Fix race condition in PipelineMainTest
| | |_|/
| |/| |
* | | |   f854b0f885 Merge pull request #9356 from retronym/ticket/11882
|\ \ \ \
| |/ / /
|/| | |
| * | | 7d81f22f4e (retronym/ticket/11882, origin/pr/9356, ticket/11882) Fix and extend Array.apply varargs optimization
```

// SKIP, 2.12.x specific
```
* | | |   fdb88812ac Merge pull request #9443 from som-snytt/issue/12315
|\ \ \ \
| * | | | 9a5fc06d52 (origin/pr/9443) Use Run reporter for wconf
|/ / / /
```

// PICK

```
* | | |   2a693ec702 Merge pull request #9355 from retronym/ticket/12213
|\ \ \ \
| |_|_|/
|/| | |
| * | | 50c2bb71c7 (retronym/ticket/12213, origin/pr/9355) Fix range position for vars with default initializer
| * | | 892a56d985 Test showing position bug with `var x = _`
| |/ /
```

// SKIP
```
* | |   2f1a80763d Merge pull request #9436 from dwijnand/2.12/mima
|\ \ \
| * | | 789fe0dc93 (origin/pr/9436) [bport] Move & truncate the MiMa filters
```

// PICK
```
* | | |   e72a873f99 Merge pull request #9315 from mkeskells/2.12.x_RedBlack_reuse_fp
|\ \ \ \
| * | | | 0678aeaa9a (origin/pr/9315) reuse common code for filter/partition for RedBlackTree based Set/Map
|/ / / /
* | | |   951c395c80 Merge pull request #9316 from mkeskells/2.12.x_RedBlack_simple
|\ \ \ \
| * | | | b990bff0e9 (origin/pr/9316) Deletion from a TreeSet/TreeMap of a key that doesn't exist should not allocate pass the tree into balance/balLeft/balRight methods so that is can be reused is unchanged
* | | | |   d613e1f5f9 Merge pull request #9046 from mkeskells/2.12.x_RedBlackTree_remove_duplication
|\ \ \ \ \
| |_|/ / /
|/| | | |
| * | | | bc742216fd (origin/pr/9046) de-duplicate code code filter and partition is common other than the supplied function
| |/ / /
```

// PICK
```
* | | |   78bf3a2d1e Merge pull request #9342 from gzm0/annotated-literal-annotation-arg
|\ \ \ \
| |/ / /
|/| | |
| * | | 2297931588 (origin/pr/9342) Allow annotated literals inside as annotation args
```

// PICK
```
* | | |   85a8814d26 Merge pull request #9369 from sjrd/remove-dead-code-patmat-array-unit
|\ \ \ \
| * | | | f39222d7bf (origin/pr/9369) Remove dead code testing for Array[Unit] after Array[AnyRef].
```

// PICK 

```
* | | | |   d057f62275 Merge pull request #9419 from koterpillar/patch-1
|\ \ \ \ \
| * | | | | a398b9b939 (origin/pr/9419) Enumeration documentation: remove nested class shadowing in "override" position
| | |_|_|/
| |/| | |
```

// PICK 
```
* | | | |   f2d2e6e6e0 Merge pull request #9426 from tanishiking/singleton-type-tree-12296
|\ \ \ \ \
| * | | | | 8a460167cf (origin/pr/9426) Ensure SingletonTypeTree#ref has symbol after typing.
| |/ / / /
* | | | |   8850b1fe0c Merge pull request #9434 from aaabramov/patch-1
|\ \ \ \ \
| * | | | | e5ff46e361 (origin/pr/9434) Fix typo in comment in Names.scala
```

// PICK

```
* | | | | |   ff03e5ad95 Merge pull request #9433 from lrytz/box-unbox-branches
|\ \ \ \ \ \
| |/ / / / /
|/| | | | |
| * | | | | 2ebf54dafe (origin/pr/9433) Fix box-unbox around branches
|/ / / / /
```

// SKIP backports and 2.12.13 release commits
```
* | | | |   09c9ef257d Merge pull request #9393 from lrytz/nowarn-nowarn
|\ \ \ \ \
| * | | | | e5580a6f61 (origin/pr/9393) [nomerge] silence warning on nowarn
* | | | | |   dd0a576ced Merge pull request #9432 from SethTisue/post-2.12.13
|\ \ \ \ \ \
| * | | | | | 457c7b44c4 (origin/pr/9432) re-STARR on 2.12.13
* | | | | | |   fd08bb447f Merge pull request #9420 from SethTisue/copyright-2021
|\ \ \ \ \ \ \
| |/ / / / / /
|/| | | | | |
| * | | | | | 906f4cd178 (origin/pr/9420) bump copyright year to 2021
| | |/ / / /
| |/| | | |
* | | | | | 60b503bcfd (tag: v2.12.13) give sonatype more time to close staging repos
* | | | | |   dfc691a5f3 Merge pull request #9430 from lrytz/build-fixes
|\ \ \ \ \ \
| |/ / / / /
|/| | | | |
| * | | | | 288926661d (origin/pr/9430) Fix release script for sbt 1.x update
|/ / / / /
```

// SKIP not applicable to groupBy in 2.13.x, the previous optimization is not yet forward ported.
```
* | | | |   b211dc4e90 Merge pull request #9376 from mkeskells/2.12.x_optimise_optimised_groupBy
|\ \ \ \ \
| * | | | | 5f0210567f (origin/pr/9376) remove some tuples
| | |_|/ /
| |/| | |
```

// SKIP backport

```
* | | | |   f5f5aae91a Merge pull request #9413 from SethTisue/travis-ci-com-2.12
|\ \ \ \ \
| * | | | | 394e6d4d0b (origin/pr/9413) [backport] move from travis-ci.org to travis-ci.com
|/ / / / /
```

// SKIP merge and immediate revert
```
* | | | |   2d69618f5e Merge pull request #9403 from lrytz/revert9392
|\ \ \ \ \
| * | | | | 4d30d48d7f (origin/pr/9403) Revert "Fix box-unbox around branches"
|/ / / / /
* | | | |   bb5e8c248b Merge pull request #9392 from lrytz/box-unbox-branches
|\ \ \ \ \
| |_|/ / /
|/| | | |
| * | | | 13a4254830 (origin/pr/9392) Fix box-unbox around branches
|/ / / /
```

// SKIP f4f6491306 was included in the forward port of that feature already. Others are backports.

```
* | | |   a90564e3ff Merge pull request #9382 from SethTisue/moar-heap
|\ \ \ \
| * | | | bd800153a3 (origin/pr/9382) [backport] backport .jvmopts from 2.13.x, sbt needs heap
|/ / / /
* | | |   471ea9d5a5 Merge pull request #9371 from retronym/2.12.x-partest-publish
|\ \ \ \
| |_|/ /
|/| | |
| * | | 58c58212c4 (retronym/2.12.x-partest-publish, origin/pr/9371) Remove non existing dependencies from scala-partest
| * | | cb865d353d Build scaladocs for partest to make sonatype happy
|/ / /
* | |   8b13dab7c8 Merge pull request #9368 from lrytz/backport-travis-wait
|\ \ \
| * | | 96db33a882 (origin/pr/9368) [backport] Avoid silence during CI, which Travis finds unacceptable
|/ / /
* | |   5c9699d081 Merge pull request #9361 from retronym/faster/class-tag-cache-property-check
|\ \ \
| * | | f4f6491306 (retronym/faster/class-tag-cache-property-check, origin/pr/9361, faster/class-tag-cache-property-check) Only look up the system prop controlling class tag caching once
| | |/
| |/|
* | |   0a3bb243ea Merge pull request #9358 from retronym/ticket/remove-pack200
|\ \ \
| |/ /
|/| |
| * | 26ceba64ac (retronym/ticket/remove-pack200, origin/pr/9358, ticket/remove-pack200) Remove pack200 ant task to make 2.12.x compilable on JDK 15
|/ /
```

// PICK

```
* |   ade1bbc42e Merge pull request #9325 from sadhen/bug_11247
|\ \
| |/
|/|
| * 398a94c235 (origin/pr/9325) Logging
| * dd5782f8e0 Catch exceptions
```

// SKIP, all backports of bug fixes in non 2.13.x code.

```
* |   aa73e74465 (mkeskells/2.12.x) Merge pull request #9335 from SethTisue/fix-array-empty-unit
|\ \
| * | 5e6676d996 (origin/pr/9335) fix regression where Array[Unit].empty caused NPE
|/ /
* |   5ee239ef63 Merge pull request #9321 from SethTisue/deprecate-arraycharsequence
|\ \
| * | b10abb48eb (origin/pr/9321) deprecate ArrayCharSequence
|/ /
* |   44be245a4a Merge pull request #9322 from SethTisue/backport-9292
|\ \
| |/
|/|
| * 6d609f1862 (origin/pr/9322) [backport] Make the CharSequence wrappers in Predef non-implicit, for JDK 15
|/
*   30c5b82ab4 Merge pull request #9311 from retronym/topic/sorted-map-set-bincompat
|\
| * b5e6d54a28 (retronym/topic/sorted-map-set-bincompat, origin/pr/9311) Avoid use of binary incompatbile super accessors in SortedMap/Set
|/
*   2166efcd2b Merge pull request #9276 from jtjeferreira/class-tag-leak
|\
| * c18fb61ac7 (origin/pr/9276) Avoid possible NPE in ClassTag cache's weak reference
| * 7fb5a1b8ce fix ClassTag cache leak
* 76f197e20c Merge pull request #9310 from retronym/backport/extra-lazy-annotation-info
* 79ae260359 (retronym/backport/extra-lazy-annotation-info, origin/pr/9310) [backport] Make annotation typechecking lazier
```

### CODE REVIEW

**Comment by SethTisue** [comment]
@retronym as per https://issues.apache.org/jira/browse/SPARK-25075, the Spark folks consider https://github.com/scala/bug/issues/12038 to be a blocker for their 2.13 upgrade, so it would be good to be extra-double-sure that this merge gets https://github.com/scala/scala/pull/9478 into 2.13.5

**Comment by retronym** [comment]
The bootstrap failure minimizes to:

```scala
package p1

class Test {

  def crash: Unit = {
    var box: AnyRef = null

    @inline def writeBox(a: AnyRef): Unit = {
      box = a
    }

    writeBox(null)
  }
}
```

```
% qscalac -d /tmp -J-Xmx1G -Yopt-trace 'p1/Test.crash' -opt:l:inline -opt-inline-from:'**/*' /tmp/HashMap.scala

/tmp/HashMap.scala:3: error: Error while emitting p1/Test
scala.tools.asm.tree.analysis.AnalyzerException: While processing p1/Test.crash
...
Caused by: scala.tools.asm.tree.analysis.AnalyzerException: Error at instruction 20: Cannot pop operand off an empty stack.
	at scala.tools.asm.tree.analysis.Analyzer.analyze(Analyzer.java:296)
	at scala.tools.nsc.backend.jvm.analysis.AsmAnalyzer.<init>(AsmAnalyzer.scala:28)
	... 54 more
Caused by: java.lang.IndexOutOfBoundsException: Cannot pop operand off an empty stack.
	at scala.tools.asm.tree.analysis.Frame.pop(Frame.java:258)
```

Here's the bytecode diff introduced by `BoxUnbox` in 2.13.x. Notice that an `ASTORE` is dropped on the grounds that is a (transitive) copy operation of the box that was being eliminated. But the following `POP` was not removed.

https://gist.github.com/retronym/d4091c9ce792cb2bab24b29e8312032c/revisions#diff-6eb5aa2af122281945d8d42cc285a04af319c927fec0b2e03ce47987c39985a2L25

image


This is related to the change in the backend around `BoxUnbox`. This doesn't crash on 2.12.x, so it could be a merging error. But it could also be a bug in 2.12.x that is hidden for some other reason.

/cc @lrytz

**Comment by retronym** [comment]
Same crash, slightly simpler bytecode. 

```scala
package p1

class Test {

  def crash: Unit = {
    var box = 0

    @inline def writeBox: Unit = {
      box = 1
    }

    writeBox
  }
}
```

https://gist.github.com/retronym/2e285db92cc4f967d68c70b532010c80/revisions#diff-6eb5aa2af122281945d8d42cc285a04af319c927fec0b2e03ce47987c39985a2L26

I don't think the inliner is essential, but so far I haven't found a repro without it.

**Comment by retronym** [comment]
I've pushed a commit with a candidate fix. I imagine this would also be needed on 2.12.x but maybe we never hit that code path in either the 2.12.x test suite or bootstrap?

**Comment by retronym** [comment]
There is a second crash with:

```scala
package p1

class AssertUtil {

  def waitForIt(terminated: => Boolean, progress: Int = 0, label: => String = "test"): Unit = {
    val limit = 5
    var n = 1
    var (dormancy, factor) = progress match {
      case 0 => (10000L, 5)
      case 1 => (250L, 4)
    }
    ()
  }
}


```

**Comment by retronym** [comment]
I think I found a fix for the second crash, https://github.com/scala/scala/pull/9392/files#diff-0ade3e36bd7d74c3cf08289672961394062ee89d088703c4ba59e06a7be0136cR344 should have used `reverseMap`.

File: project/MimaFilters.scala [hunk 1]
```diff
@@ -30,6 +30,13 @@ object MimaFilters extends AutoPlugin {
 
     // #9487
     ProblemFilters.exclude[MissingClassProblem]("scala.reflect.ClassTag$cache$"),
+
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.redWithRight"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.redWithLeftRight"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.blackWithLeftRight"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree.partitionKeys"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree.partitionKeys"),
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree#Tree.redWithLeft"),
```
Comment (dwijnand) [inline project/MimaFilters.scala:37]: These are already fixed in 2.12.x, btw (in a more general way): https://github.com/scala/scala/pull/9507/files, so no need to fast-back-track-port.

**Review by lrytz — APPROVED** [review]
...whew 😰 thank you!

**Comment by SethTisue** [comment]
marked "release-notes" because 2.12.14 isn't out yet, so the 2.13.5 notes may need to hit some highlights

**Pull request merged**